src/models/linear_network.py

```diff
@@ -0,0 +1,196 @@
+        self.model: LinearModel = LinearModel(input_size=self.input_size,
+                                              layer_sizes=self.layer_sizes,
+                                              dropout=self.dropout)
+
+        optimizer = torch.optim.Adam([pam for pam in self.model.parameters()],
+                                     lr=learning_rate)
+
+        for epoch in range(num_epochs):
+            train_rmse = []
+            self.model.train()
+            for x, y in train_dataloader:
+                for x_batch, y_batch in chunk_array(x, y, batch_size, shuffle=True):
+                    optimizer.zero_grad()
+                    pred = self.model(x_batch)
+                    loss = F.smooth_l1_loss(pred, y_batch)
+                    loss.backward()
+                    optimizer.step()
+
+                    train_rmse.append(loss.item())
+
+            if early_stopping is not None:
+                self.model.eval()
+                val_rmse = []
+                with torch.no_grad():
+                    for x, y in val_dataloader:
+                        val_pred_y = self.model(x)
+                        val_loss = F.mse_loss(val_pred_y, y)
+
+                        val_rmse.append(val_loss.item())
+
+            print(f'Epoch {epoch + 1}, train RMSE: {np.mean(train_rmse)}')
+
+            if early_stopping is not None:
+                epoch_val_rmse = np.mean(val_rmse)
+                print(f'Val RMSE: {epoch_val_rmse}')
+                if epoch_val_rmse < best_val_score:
+                    batches_without_improvement = 0
+                    best_val_score = epoch_val_rmse
+                else:
+                    batches_without_improvement += 1
+                    if batches_without_improvement == early_stopping:
+                        print('Early stopping!')
+                        return None
+
+    def predict(self) -> Tuple[Dict[str, Dict[str, np.ndarray]], Dict[str, np.ndarray]]:
+        test_arrays_loader = DataLoader(data_path=self.data_path, batch_file_size=self.batch_size,
+                                        shuffle_data=False, mode='test', to_tensor=True)
+
+        preds_dict: Dict[str, np.ndarray] = {}
+        test_arrays_dict: Dict[str, Dict[str, np.ndarray]] = {}
+
+        if self.model is None:
+            self.train()
+            self.model: LinearModel
+
+        self.model.eval()
+        with torch.no_grad():
+            for dict in test_arrays_loader:
+                for key, val in dict.items():
+                    preds = self.model(val.x)
+                    preds_dict[key] = preds.numpy()
+                    test_arrays_dict[key] = {'y': val.y.numpy(), 'latlons': val.latlons}
+
+        return test_arrays_dict, preds_dict
+
+
+class LinearModel(nn.Module):
+
+    def __init__(self, input_size, layer_sizes, dropout):
+        super().__init__()
+        layer_sizes.insert(0, input_size)
```

Comment:
When i run the code it seems that the initial layer for me is missing. What have I forgotten?

```
x_ref, _ = next(iter(train_dataloader))
input_size = x_ref.contiguous().view(x_ref.shape[0], -1).shape
layer_sizes = [100]
dropout = 0.25

lm = LinearModel(input_size=input_size,
                layer_sizes=l.layer_sizes,
                dropout=dropout)
lm.init_weights()
learning_rate = 1e-3
optimizer = torch.optim.Adam(
    [pam for pam in lm.parameters()],
    lr=learning_rate
)

# ------------------------------------------------------------------------------
# TRAINING
# ------------------------------------------------------------------------------

num_epochs = 1
early_stopping = None
batch_size = 5
learning_rate = 1e-3

train_rmse = []
# set in training mode
lm.train()


In [53]: for x, y in train_dataloader:
    ...:     for x_batch, y_batch in chunk_array(x, y, batch_size):
    ...:         optimizer.zero_grad()
    ...:         pred = lm(x_batch)
    ...:         loss = F.smooth_l1_loss(pred, y_batch)
    ...:         loss.backward()
    ...:         optimizer.step()
    ...:         train_rmse.append(loss.item())
    ...:
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-53-8542f66a653f> in <module>
      2     for x_batch, y_batch in chunk_array(x, y, batch_size):
      3         optimizer.zero_grad()
----> 4         pred = lm(x_batch)
      5         loss = F.smooth_l1_loss(pred, y_batch)
      6         loss.backward()

~/miniconda3/envs/esowc-drought/lib/python3.7/site-packages/torch/nn/modules/module.py in __call__(self, *input, **kwargs)
    491             result = self._slow_forward(*input, **kwargs)
    492         else:
--> 493             result = self.forward(*input, **kwargs)
    494         for hook in self._forward_hooks.values():
    495             hook_result = hook(self, input, result)

<ipython-input-6-ebc273557471> in forward(self, x)
     25             x = layer(x)
     26
---> 27         return self.final_dense(x)
     28
     29

~/miniconda3/envs/esowc-drought/lib/python3.7/site-packages/torch/nn/modules/module.py in __call__(self, *input, **kwargs)
    491             result = self._slow_forward(*input, **kwargs)
    492         else:
--> 493             result = self.forward(*input, **kwargs)
    494         for hook in self._forward_hooks.values():
    495             hook_result = hook(self, input, result)

~/miniconda3/envs/esowc-drought/lib/python3.7/site-packages/torch/nn/modules/linear.py in forward(self, input)
     90     @weak_script_method
     91     def forward(self, input):
---> 92         return F.linear(input, self.weight, self.bias)
     93
     94     def extra_repr(self):

~/miniconda3/envs/esowc-drought/lib/python3.7/site-packages/torch/nn/functional.py in linear(input, weight, bias)
   1404     if input.dim() == 2 and bias is not None:
   1405         # fused op is marginally faster
-> 1406         ret = torch.addmm(bias, input, weight.t())
   1407     else:
   1408         output = input.matmul(weight.t())

RuntimeError: size mismatch, m1: [5 x 55], m2: [100 x 1] at /opt/conda/conda-bld/pytorch_1556653114079/work/aten/src/TH/generic/THTensorMath.cpp:961
```

Reply:
Ah, the mistake is `input_size`, which is a float, not a `torch.Size`; see above.

Reply:
You are my hero.